src/Dialect/ONNX/ONNXOpsHelper.cpp

```diff
@@ -45,6 +46,110 @@ AffineMap getConvDimMap(Builder &builder, bool ceilMode) {
   return AffineMap::get(1, 4, {dimExp});
 }
 
+/// IndexExprs to compute the start and end indices of the convolution/pooling
+/// window.
+///
+/// The conv/pooling window can be smaller than the kernel when slicing it over
+/// the border edges. Thus, we will compute the start and end indices for
+/// each window dimension as follows.
+///   firstValidH = ceil(float(ptH / dH)) * dH - ptH
+///   startH = max(firstValidH, ho * sH - ptH)
+///   endH = min(H, ho * sH + (kH - 1) * dH  + 1 - pbH)
+///
+/// Full conv/pooling window can be reconstructed by:
+///   hDim = round(float(endH - startH) / float(dH))
+//
+/// We also want to compute how the window is smaller than the kernel.
+///   kernelOffset = min(0, ho * sH - ptH)
+///
+/// How to derive 'firstValidH':
+///   When dilation is non-unit, the first valid pixel to apply conv/pooling on
+///   will not be the 0-th pixel, but rather the smallest integer n to make
+///   '-pH + n * dH' greater than or equal to 0, where pH and dH are pad
+///   and dilation along axis H. We derive what is this smallest n:
+///   -pH + n * dH >= 0
+///         n * dH >= pH
+///              n >= pH/dH
+///   thus n = ceil(pH/dH)
+///   thus the first valid pixel location is 'ceil(pH / dH) * dH- pH'.
+///
+/// This function returns {startH, endH, kernelOffset}.
+
+std::vector<IndexExpr> getIndexExprsForConvWindow(IndexExprContext &context,
+    SmallVectorImpl<IndexExpr> &inputExprs, bool ceilMode, bool isDilated) {
+  assert(inputExprs.size() == 6 && "Not enought inputs");
+  IndexExpr windowStartExpr, windowEndExpr, kernelOffsetExpr;
+  IndexExpr outputIndex = inputExprs[0];
+  IndexExpr inputDim = inputExprs[1];
+  IndexExpr kernelDim = inputExprs[2];
+  IndexExpr padTopDim = inputExprs[3];
+  IndexExpr strideDim = inputExprs[4];
+  IndexExpr dilationDim = inputExprs[5];
+
+  IndexExpr start1 = (padTopDim).ceilDiv(dilationDim) * dilationDim - padTopDim;
+  IndexExpr start2 = outputIndex * strideDim - padTopDim;
+  IndexExpr end1 = inputDim;
+  IndexExpr end2 =
+      outputIndex * strideDim + (kernelDim - 1) * dilationDim + 1 - padTopDim;
+
+  // windowStartExpr
+  SmallVector<mlir::IndexExpr, 2> startExprs = {start1, start2};
+  windowStartExpr = IndexExpr::max(startExprs);
+  // windowEndExpr
+  SmallVector<mlir::IndexExpr, 2> endExprs = {end1, end2};
+  windowEndExpr = IndexExpr::min(endExprs);
+  // kernelOffsetExpr
+  SmallVector<mlir::IndexExpr, 2> kernelExprs = {
+      context.createLiteralIndex(0), start2};
+  kernelOffsetExpr = IndexExpr::min(kernelExprs);
+
+  return std::vector<IndexExpr>{
+      windowStartExpr, windowEndExpr, kernelOffsetExpr};
+}
+
+/// The conv/pooling window can be smaller than the kernel when slicing it over
+/// the border edges. This function returns an AffineMap to compute the size of
+/// one edge of the window.
+AffineMap getWindowAffineMap(Builder &builder, bool ceilMode, bool isDilated) {
```

Comment:
@AlexandreEichenberger FYI, this function produces an affine map that will be used as the upperbound for krnl.iterate. It's great if we can use IndexExpr here.